### PR TITLE
Cache Partitioning Predicate from Scalar Expr

### PR DESCRIPTION
**Pull request by bhuvnesh2703**
Given a ScalarCmp Expression, we need not derive the scalar conditions
which are based on the partitioning predicate everytime. They should be
calculated once for a ScalarCmp expression and cached. This saves
optimization time.

Input Expr:
```
        +--CScalarBoolOp (EboolopAnd)
        |--CScalarArrayCmp Any (=)
        |  |--CScalarIdent "risk_set_row_id" (2)
        |  +--CScalarArray: {eleMDId: (23,1.0), arrayMDId: (1007,1.0) CScalarConst (32) CScalarConst (33) CScalarConst (43) CScalarConst (9) CScalarConst (10) CScalarConst (15) CScalarConst (16) CScalarConst (36) CScalarConst (11) CScalarConst (50) CScalarConst (46) CScalarConst (356) CScalarConst (468) CScalarConst (42)}
        +--CScalarCmp (=)
        |--CScalarIdent "value_date" (0)
        +--CScalarConst (559094400000000.000)
```

Let's say the partition key is on value_date, then the extracted
predicate is as below and is cached:
Output Expr:
```
        +--CScalarCmp (=)
        |--CScalarIdent "value_date" (0)
        +--CScalarConst (559094400000000.000)
```

This brings down the optimization time of the north star query from 47.x sec to 42.y (around 10.6% improvement)
#of total request 5902
#Retrieved from Cached: 5794

### CODE REVIEW

File: libgpopt/src/operators/CLogicalSelect.cpp [hunk 1]
```diff
@@ -237,20 +253,38 @@ CLogicalSelect::PexprPartPred
...
-												(*pdrgppartkeys)[ul]->Pdrgpdrgpcr(),
-												NULL, //pcrsAllowedRefs
-												true //fUseConstraints
-												);
+		// insert the scalar expression and the corresponding partitioning predicate
+		// in the hashmap
+		pexprPredOnPartKey->AddRef();
+		pexprScalar->AddRef();
+		m_hmPexprPred->FInsert(pexprScalar, pexprPredOnPartKey);
 	}
```
Comment (hsyuan) [inline libgpopt/src/operators/CLogicalSelect.cpp]: The logic of this function needs to be refined.
```cpp
do cache lookup
if exist in chace
   then return cached object

extract predicates
insert into cache
return extracted predicates
```

File: libgpopt/include/gpopt/operators/CLogicalSelect.h [hunk 2]
```diff
@@ -34,6 +34,11 @@ namespace gpopt
 			// private copy ctor
 			CLogicalSelect(const CLogicalSelect &);
 
+			typedef CHashMap<CExpression, CExpression, CExpression::UlHash, CUtils::FEqual,
+				CleanupRelease<CExpression>, CleanupRelease<CExpression> > HMPexprPred;
```
Comment (hsyuan) [inline libgpopt/include/gpopt/operators/CLogicalSelect.h]: better put this before the class definition

File: libgpopt/src/operators/CLogicalSelect.cpp [hunk 3]
```diff
@@ -237,20 +253,38 @@ CLogicalSelect::PexprPartPred
...
+		for (ULONG ul = 0; NULL == pexprPredOnPartKey && ul < ulKeySets; ul++)
+		{
+			pexprPredOnPartKey = CPredicateUtils::PexprExtractPredicatesOnPartKeys
+			(
+			 pmp,
+			 pexprScalar,
+			 (*pdrgppartkeys)[ul]->Pdrgpdrgpcr(),
+			 NULL, //pcrsAllowedRefs
+			 true //fUseConstraints
+			 );
```
Comment (hsyuan) [inline libgpopt/src/operators/CLogicalSelect.cpp]: let's indent above lines.

File: libgpopt/src/operators/CLogicalSelect.cpp [hunk 4]
```diff
@@ -237,20 +253,38 @@ CLogicalSelect::PexprPartPred
...
+													NULL, //pcrsAllowedRefs
+													true //fUseConstraints
+													);
+		}
+	}
 
-	DrgPpartkeys *pdrgppartkeys = ppartinfo->Pdrgppartkeys(0 /*ulPos*/);
-	const ULONG ulKeySets = pdrgppartkeys->UlLength();
-	for (ULONG ul = 0; NULL == pexprPredOnPartKey && ul < ulKeySets; ul++)
+	if (pexprPredOnPartKey != NULL)
```
[the remark this answers is not in the document]
Comment (bhuvnesh2703) [reply libgpopt/src/operators/CLogicalSelect.cpp]: makes sense. done.

File: libgpopt/include/gpopt/operators/CLogicalSelect.h [hunk 5]
```diff
@@ -34,6 +37,8 @@ namespace gpopt
 			// private copy ctor
 			CLogicalSelect(const CLogicalSelect &);
 
+			HMPexprPartPred *m_hmPexprPartPred;
```
Comment (oarap) [inline libgpopt/include/gpopt/operators/CLogicalSelect.h]: `m_phmPexprPartPred`

**Pull request merged**